huggle/login.cpp

```diff
@@ -1269,42 +1269,71 @@ bool Login::ProcessOutput(WikiSite *site)
     ApiQuery *query = this->LoginQueries[site];
     // Check what the result was
     ApiQueryResult *result = query->GetApiQueryResult();
-    ApiQueryResultNode *ln = result->GetNode("login");
-    QString result_code = ln->GetAttribute("result");
-    QString reason = ln->GetAttribute("reason");
-    if (result_code.isEmpty())
+    ApiQueryResultNode *ln = result->GetNode("clientlogin");
+    QString status = ln->GetAttribute("status");
+    if (status.isEmpty())
     {
         this->DisplayError(_l("api.php-invalid-response"));
         return false;
     }
-    if (result_code == "Success")
+
+    if (status == "PASS")
         return true;
-    if (result_code == "EmptyPass")
-    {
-        this->DisplayError(_l("login-password-empty"));
+    if (status == "UI") {
+        // Need a user interaction like captacha or 2FA
+        //QString v_id = ln->ChildNodes.at(0)->GetAttribute("id", "unknown");
+        //if (v_id == "TOTPAuthenticationRequest"){
+        if (true){
+            // 2FA is requierd (TOTP code needed)
+            QString totp = QInputDialog::getText(this, "Two factor authentification", "Please enter the 2FA code from your device:");
+            query = new ApiQuery(ClientLogin, site);
+            //query->HiddenQuery = true;
+            query->IncRef();
+            query->Parameters = "username=" + QUrl::toPercentEncoding(hcfg->SystemConfig_BotLogin)
+                        + "&password=" + QUrl::toPercentEncoding(hcfg->TemporaryConfig_Password)
+                        + "&OATHToken=" + totp + "&loginreturnurl=http://example.com/&rememberMe=1&logintoken=" + QUrl::toPercentEncoding(this->Tokens[site]);
+            query->UsingPOST = true;
+            query->Process();
+            ApiQueryResult *result = query->GetApiQueryResult();
+            ApiQueryResultNode *ln = result->GetNode("clientlogin");
```

Comment:
We need to execute this second request, but I can't finish it correctly without spending it with other connection requests.